api/controller/update/controller.go

```
@@ -104,20 +104,25 @@ func (u *Controller) updateDeployment(c *api.Cluster, yamlFiles []string, master
 
 func (u *Controller) updateEtcdCluster(c *api.Cluster, yamlFiles []string, masterVersion *api.MasterVersion, nextPhase api.MasterUpdatePhase) (*api.Cluster, error) {
 	for _, yamlFile := range yamlFiles {
-		etcd, err := resources.LoadEtcdClusterFile(masterVersion, u.MasterResourcesPath, yamlFile)
+		newEtcd, err := resources.LoadEtcdClusterFile(masterVersion, u.MasterResourcesPath, yamlFile)
 		if err != nil {
 			return nil, err
 		}
 
-		_, err = u.EtcdClusterClient.Cluster(fmt.Sprintf("cluster-%s", c.Metadata.Name)).Update(etcd)
+		oldEtcd, err := u.EtcdClusterClient.Cluster(fmt.Sprintf("cluster-%s", c.Metadata.Name)).Get("etcd-cluster")
+		if err == nil {
+			oldEtcd.Spec.Version = newEtcd.Spec.Version
+		}
+		_, err = u.EtcdClusterClient.Cluster(fmt.Sprintf("cluster-%s", c.Metadata.Name)).Update(oldEtcd)
+
 		if errors.IsNotFound(err) {
```

You are loading the etcd cluster, set the version, and update it.
And only when you update it, you check if the cluster actually exists?

This seems a bit odd, is there any reason for it?

IMO, we should not create a new cluster if no old one was found. If we could not find a cluster, we should throw an error. Something is very wrong then